### PR TITLE
(fix): Update Kagane

### PR DESCRIPTION
**Pull request by Zephyris-Pro**
Hey, I’ve fixed Kagane.

The `FetchMangas` function is probably not optimized at all and could potentially cause IP timeouts for some users (it doesn’t for me).

If it’s possible to optimize it or reduce the number of requests, that would be great, especially since not all mangas are being fetched. I’ve got roughly ~7100.

The DRM part has also been updated. For more information, you can check the Discord.

### CODE REVIEW

**Comment by MikeZeDev** [comment]
I dont think we can properly review obfuscated code 😔

**Comment by Zephyris-Pro** [comment]
It might be more readable now 😅

File: web/src/engine/websites/Kagane.DRM.js [hunk 1]
```diff
@@ -1,2 +1,367 @@
...
+  optional bytes token = 2;
+  repeated NameValue client_info = 3;
+  optional bytes provider_client_token = 4;
+  optional uint32 license_counter = 5;
+  optional ClientCapabilities client_capabilities = 6;
+}
+
+message EncryptedClientIdentification {
+  optional string service_id = 1;
+  optional bytes service_certificate_serial_number = 2;
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.DRM.js]: Make a separate Kagane.proto file and import the file as raw.

Check other websites like ComicFuz

`import protoTypes from './ComicFuz.proto?raw';`

File: web/src/engine/websites/Kagane.ts [hunk 2]
```diff
@@ -1,78 +1,163 @@
...
+
+        // All meaningful sort strategies to maximize coverage past the 1000-result cap
+        const sortStrategies = [
+            'avg_views_today,desc',
+            'avg_views_today,asc',
+            'avg_views_week,desc',
+            'avg_views_week,asc',
+            'avg_views_month,desc',
+            'avg_views_month,asc',
+            'total_views,desc',
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: I remember the old Comix limiting results to 1000. I think we should respect that . Just chose a sorting . User can paste manga url in case its missing.

File: web/src/engine/websites/Kagane.ts [hunk 3]
```diff
@@ -1,78 +1,163 @@
...
+        );
+        tokenEndpoint.searchParams.set('is_datasaver', 'false');
+
+        const pages = await this.#drm.CreateImageURL(
+            tokenEndpoint.toString(),
+            mangaId,
+            chapterId
+        );
+
+        return pages.map(({ link, parameters }) =>
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: keep that on one line

File: web/src/engine/websites/Kagane.ts [hunk 4]
```diff
@@ -1,78 +1,163 @@
...
+            new Page<PageParameters>(this, chapter, link, parameters)
+        );
     }
 
-    public override async FetchImage(page: Page<PageParameters>, priority: Priority, signal: AbortSignal): Promise<Blob> {
+    public override async FetchImage(
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: keep that on one line

File: web/src/engine/websites/Kagane.ts [hunk 5]
```diff
@@ -1,78 +1,163 @@
...
-        const uri = new URL(`./books/${manga.Identifier}`, this.#apiURL);
-        const { content } = await FetchJSON<APIChapters>(new Request(uri));
-        return content.map(({id, title}) => new Chapter(this, manga, id, title));
     }
-
     public override async FetchPages(chapter: Chapter): Promise<Page<PageParameters>[]> {
-        const uri = new URL(`./books/${chapter.Parent.Identifier}/file/${chapter.Identifier}`, this.#apiURL);
-        const pages = await this.#drm.CreateImageLinks(uri, chapter.Parent.Identifier, chapter.Identifier);
-        return pages.map(({ link, parameters })=> new Page<PageParameters>(this, chapter, link, parameters));
+        const mangaId = chapter.Parent.Identifier;
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: those two variable have no real use

File: web/src/engine/websites/Kagane.ts [hunk 6]
```diff
@@ -1,78 +1,163 @@
...
     }
 
-    public override async FetchImage(page: Page<PageParameters>, priority: Priority, signal: AbortSignal): Promise<Blob> {
+    public override async FetchImage(
+        page: Page<PageParameters>,
+        priority: Priority,
+        signal: AbortSignal
+    ): Promise<Blob> {
         const bytes = await this.imageTaskPool.Add(async () => {
-            const response = await Fetch(new Request(page.Link, { signal: signal }));
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts:72]:
```ts
        const blob = await Common.FetchImageAjax.call(this, page, priority, signal);
        return this.#drm.DecryptImage(await blob.arrayBuffer());
```

File: web/src/engine/websites/Kagane.ts [hunk 7]
```diff
@@ -1,78 +1,163 @@
...
+        const mangaId = chapter.Parent.Identifier;
+        const chapterId = chapter.Identifier;
+
+        const tokenEndpoint = new URL(
+            `./books/${chapterId}`,
+            this.#apiURL
+        );
+        tokenEndpoint.searchParams.set('is_datasaver', 'false');
+
+        const pages = await this.#drm.CreateImageURL(
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: keep that on one line

File: web/src/engine/websites/Kagane.ts [hunk 8]
```diff
@@ -1,78 +1,163 @@
...
     }
 
-    /* ⚠️ Disabled for now due to high risk of IP ban (429 => Too many Requests)
-    // TODO: Consider adding this to the website list provider repository
-    public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
+    public override async FetchChapters(manga: Manga): Promise<Chapter[]> {
+        const uri = new URL(`./series/${manga.Identifier}`, this.#apiURL);
+        const { series_books } = await FetchJSON<APIChapters>(new Request(uri));
+
+        return series_books.map(book => {
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]:
```ts
        return series_books.map(({ book_id : bookId, title,chapter_no )=>  new Chapter(this, manga, bookId, title ?? `Chapter ${chapter_no}`));
```

File: web/src/engine/websites/Kagane.ts [hunk 9]
```diff
@@ -1,78 +1,163 @@
...
 
     readonly #drm = new DRMProvider();
-    readonly #apiURL = 'https://api.kagane.org/api/v1/';
+
+    readonly #apiURL = 'https://yuzuki.kagane.org/api/v2/';
 
     public constructor() {
-        super('kagane', 'Kagane', 'https://kagane.org', Tags.Media.Manga, Tags.Media.Manhwa, Tags.Media.Manhua, Tags.Language.English, Tags.Source.Aggregator, ...Tags.Rating.toArray());
+        super(
+            'kagane', 'Kagane', 'https://kagane.org',
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: only one line pls

File: web/src/engine/websites/Kagane.ts [hunk 10]
```diff
@@ -1,78 +1,163 @@
...
+            Tags.Media.Manga, Tags.Media.Manhwa, Tags.Media.Manhua,
+            Tags.Language.English, Tags.Source.Aggregator,
+            ...Tags.Rating.toArray()
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: Rating to array?

File: web/src/engine/websites/Kagane.ts [hunk 11]
```diff
@@ -1,78 +1,163 @@
...
     }
-
     public override async FetchPages(chapter: Chapter): Promise<Page<PageParameters>[]> {
-        const uri = new URL(`./books/${chapter.Parent.Identifier}/file/${chapter.Identifier}`, this.#apiURL);
-        const pages = await this.#drm.CreateImageLinks(uri, chapter.Parent.Identifier, chapter.Identifier);
-        return pages.map(({ link, parameters })=> new Page<PageParameters>(this, chapter, link, parameters));
+        const mangaId = chapter.Parent.Identifier;
+        const chapterId = chapter.Identifier;
+
+        const tokenEndpoint = new URL(
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.ts]: keep that on one line

File: web/src/engine/websites/Kagane.DRM.d.ts [hunk 12]
```diff
@@ -1,16 +1,27 @@
...
+    parameters: PageParameters;
+};
+
 export declare class DRMProvider {
     #private;
     constructor();
-    CreateImageLinks(chapterURL: URL, mangaIdentifier: string, chapterIdentifier: string): Promise<Page[]>;
-    DecryptImage(bytes: ArrayBuffer, parameters: PageParameters): Promise<Blob>;
+    CreateImageURL(
+        tokenEndpointUrl: string,
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.DRM.d.ts]: keep that on one line

File: web/src/engine/websites/Kagane.DRM.d.ts [hunk 13]
```diff
@@ -1,16 +1,27 @@
...
+        bookId: string,
+        assetId: string
+    ): Promise<Page[]>;
+
+    DecryptImage(
```
Comment (MikeZeDev) [inline web/src/engine/websites/Kagane.DRM.d.ts]: keep that on one line

**Review by MikeZeDev — CHANGES_REQUESTED** [review]
A couple of cosmetic  changes requested, cant talk about the rest

File: web/src/engine/websites/Kagane.ts (hunk 5 above)
Comment (Zephyris-Pro) [reply web/src/engine/websites/Kagane.ts]: `chapterId` is reused 2 times

File: web/src/engine/websites/Kagane.ts (hunk 10 above)
Comment (Zephyris-Pro) [reply web/src/engine/websites/Kagane.ts]: Tbh, I don't know why this was here in the first place 😅, so I just left it here.

Comment (Zephyris-Pro) [reply web/src/engine/websites/Kagane.ts]: All the tags work for Kagane.

File: web/src/engine/websites/Kagane.ts (hunk 5 above)
Comment (MikeZeDev) [reply web/src/engine/websites/Kagane.ts]: What i mean is you dont create a duplicate of chapter.Parent.Identifier or chapter.Identifier, use them directly

**Comment by Zephyris-Pro** [comment]
@MikeZeDev, is there anything to change ?

**Comment by MikeZeDev** [comment]
> @MikeZeDev, is there anything to change ?

Idk anything about Kagane and Widevine, i gave same small observations about the code. I am no confident reviewing that besides those 😔

**Comment by 27136773** [comment]
Please check the Google Widevine ToS if it is allowed to host and share proprietary code such as the Widevine `proto` types

File: web/src/engine/websites/Kagane.ts [hunk 14]
```diff
@@ -29,50 +32,47 @@ export default class extends DecoratableMangaScraper {
         return icon;
     }
 
-    /* ⚠️ Disabled for now due to high risk of IP ban (429 => Too many Requests)
-    // TODO: Consider adding this to the website list provider repository
     public override async FetchMangas(provider: MangaPlugin): Promise<Manga[]> {
```
Comment (27136773) [inline web/src/engine/websites/Kagane.ts:35]: Leave this method disabled, it will be added to the manga-list repository some time in the distant future

File: web/src/engine/websites/Kagane.ts [hunk 15]
```diff
@@ -29,50 +32,47 @@ export default class extends DecoratableMangaScraper {
...
+        const body = JSON.stringify({ content_rating: ['Safe', 'Suggestive', 'Erotica'] }); // Removed 'Pornographic' as it was always returning that kind of content.
+        return await Array.fromAsync(async function* () {
+            for (let page = 0; ; page++) {
+                uri.searchParams.set('page', String(page));
+                const data = await FetchJSON<APIMangas>(new Request(uri, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body }));
+                if (!data.content.length) break;
+                for (const { series_id, title } of data.content) {
+                    yield new Manga(this, provider, series_id, title);
+                }
+                if (page + 1 >= data.total_pages) break;
```
Comment (27136773) [inline web/src/engine/websites/Kagane.ts:54]: What was wrong with the original condition?
Consider using `mangas.length > 0 ? yield* mangas : run = false;` since it is consistent with the approach how it is done in other websites.
In case the last page is repeated, consider using the `isMissingLastItemFrom` array extension method.

File: web/src/engine/websites/Kagane.ts [hunk 16]
```diff
@@ -29,50 +32,47 @@ export default class extends DecoratableMangaScraper {
...
-                    method: 'POST',
-                    headers: { 'Content-Type': 'application/json' },
-                    body: JSON.stringify({ sources: [], content_rating: ['safe', 'suggestive', 'erotica', 'pornographic']}),
-                }));
-                const mangas = content.map(({ id, name }) => new Manga(this, provider, id, name));
-                mangas.length > 0 ? yield* mangas : run = false;
-                await Delay(2500);
+        const uri = new URL('./search/series', this.#apiURL);
+        uri.search = new URLSearchParams({ size: '100', sort: 'updated_at,desc' }).toString();
+        const body = JSON.stringify({ content_rating: ['Safe', 'Suggestive', 'Erotica'] }); // Removed 'Pornographic' as it was always returning that kind of content.
```
Comment (27136773) [inline web/src/engine/websites/Kagane.ts:45]: Did the website really changed the search terms to pascal casing?

File: web/src/engine/websites/Kagane.ts [hunk 17]
```diff
@@ -29,50 +32,47 @@ export default class extends DecoratableMangaScraper {
...
+        return await Array.fromAsync(async function* () {
+            for (let page = 0; ; page++) {
+                uri.searchParams.set('page', String(page));
+                const data = await FetchJSON<APIMangas>(new Request(uri, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body }));
+                if (!data.content.length) break;
+                for (const { series_id, title } of data.content) {
```
Comment (27136773) [inline web/src/engine/websites/Kagane.ts:51]: Use `.map` and `yield*` instead of loop

File: web/src/engine/websites/Kagane.ts [hunk 18]
```diff
@@ -29,50 +32,47 @@ export default class extends DecoratableMangaScraper {
...
+        const uri = new URL(`./series/${manga.Identifier}`, this.#apiURL);
+        const { series_books } = await FetchJSON<APIChapters>(new Request(uri));
+        return series_books.map(({ book_id: bookId, title, chapter_no: chapterNo }) => new Chapter(this, manga, bookId, title ?? `Chapter ${chapterNo}`));
     }
 
     public override async FetchPages(chapter: Chapter): Promise<Page<PageParameters>[]> {
-        const uri = new URL(`./books/${chapter.Parent.Identifier}/file/${chapter.Identifier}`, this.#apiURL);
-        const pages = await this.#drm.CreateImageLinks(uri, chapter.Parent.Identifier, chapter.Identifier);
-        return pages.map(({ link, parameters })=> new Page<PageParameters>(this, chapter, link, parameters));
+        const chapterId = chapter.Identifier;
```
Comment (27136773) [inline web/src/engine/websites/Kagane.ts:67]: Why introducing this intermediate variable `chapterId`, there seems to be no benefit?

File: web/src/engine/websites/Kagane.DRM.js [hunk 19]
```diff
@@ -1,2 +1,148 @@
...
+    async #CreateIntegrityChallenge() {
+        const { token, exp } = await FetchJSON(
+            new Request(`https://kagane.org/api/integrity`, {
+                method: "POST",
+            })
+        );
+        return { token, exp };
+    }
+
+    async CreateImageURL(tokenEndpointUrl, bookId, assetId, isDataSaver = false) {
```
Comment (27136773) [inline web/src/engine/websites/Kagane.DRM.js:110]: No need for a data-saver option, the software is supposed to always deliver the best images

File: web/src/engine/websites/Kagane.DRM.js [hunk 20]
```diff
@@ -1,2 +1,148 @@
...
+    }
+
+    async #CreateMediaLicenseChallenge(psshBuffer) {
+        const psshBytes = new Uint8Array(psshBuffer).slice(32);
+
+        const licenseRequest = this.#protoLicenseRequest.create({
+            type: 1,
+            protocol_version: 22,
+            request_time: Date.now() / 1000 >>> 0,
+            key_control_nonce: Math.random() * 0xFFFFFFFF >>> 0,
```
Comment (27136773) [inline web/src/engine/websites/Kagane.DRM.js:75]: The bit-mask seems slightly incorrect, it may never produce `4'294'967'295`

File: web/src/engine/websites/Kagane.DRM.js [hunk 21]
```diff
@@ -1,2 +1,148 @@
...
\ No newline at end of file
+import protobuf from 'protobufjs';
+import protoTypes from './Kagane.proto?raw';
+import { FetchJSON } from "../platform/FetchProvider";
+import { GetBase64FromBytes } from "../BufferEncoder";
+import { RandomBytes } from "../Random";
+
+
+// Hardcoded Widevine SystemID for the PSSH box: edef8ba979d64acea3c827dcd51d21ed
+// Available in https://kagane.org/_next/static/chunks/common-e6bef285f8b57c56.js
```
Comment (27136773) [inline web/src/engine/websites/Kagane.DRM.js:10]: Consider removing this comment, this is [not unique](https://www.google.com/search?q=edef8ba979d64acea3c827dcd51d21ed) to Kagane

**Comment by 27136773** [comment]
Avoid using double quotes typescript/javascript, use single quotes instead

**Comment by 27136773** [comment]
Migrated in 4e3dc18e
